### DISCUSSION TITLE
STM32 PWM Input Capture Issue

### DISCUSSION THREAD

**Discussion by lainslie** [zephyrproject-rtos/zephyr#53514] [Q&A]
Hi there,

I am using a custom STM32F103C8T6 board. Part of the design is to determine a fans speed by its cyclic period, an example signal has been attached:

![image](https://user-images.githubusercontent.com/58678981/210711611-747e6ba2-53be-4383-9375-702f772528c0.png)

This fan has been connected to TIM3 ch2 PA7. See the device tree setup for this device:

![image](https://user-images.githubusercontent.com/58678981/210712412-59e98623-7ed6-4205-87b7-76881000aff5.png)

Within my proj.conf, I have enabled CONFIG_PWM and CONFIG_PWM_CAPTURE. 

I am currently experiencing an issue that occurs when I initialise the PWM capture for that pin. The PWM output of the fan seems to be 'driven' to low (0 voltage), seen by the logic analyser, which means I am unable to read a value from this pin, since no cycles are occurring. I'm unsure of what is causing this issue, the only idea in my head is that the pin by default is configured as an output pin, since PWM is typical an output signal, and is not reconfigured as an input pin when pwm_capture_usec, pwm_pin_configure_capture or pwm_pin_enable_capture is called. Any help would be greatly appreciated thanks

**Comment by erwango** [comment]
The GPIO mode used to configure a pin as a TIMx_CH, is "alternate function" which different from output/input GPIO.
What you can change is the drive (Push-pull/Open-Drain), but I'm not sure it can have an impact in your case.
Default is push-pull, you might want to use open-drain:
```
&tim3_ch2_pwm_pa7 {
    drive-open-drain;
};
```

Otherwise, I'd suggest to check:
- behavior of the pin in PWM mode
- possible pin conflicts

**Reply by lainslie** [reply]
Hi mate thanks for your suggestion. As you suspected, the "drive-open-drain" did not impact the results, and just as a sanity check for my own sake, I set the tim3_ch2_pa7 pinmux to all available remap options, without any success. When the pin in put in pwm mode, the signal is outputted as expected. I've also disabled all other peripherals to ensure that there are no pin conflicts. Unfortunately the issue is STILL occurring and I am genuinely out of ideas. 

Side note:
- When pwm3 is disabled, the signal is viewed correctly by the logic analyser. 

Not sure if this is a zephyr issue or if anyone has suggestions before it can be moved to an issue. Cheers

**Reply by lainslie** [reply]
I have also tried with pwm4, so it is not a specific issue with pwm3 or that pin

**Reply by erwango** [reply]
I suspect this is a PWM Capture - F1 specific issue

**Reply by lainslie** [reply]
How should I proceed ? Making this an issue or make a new issue ?

**Reply by erwango** [reply]
Let's make it an issue
